src/bpfprogram.cpp

```diff
@@ -35,16 +75,195 @@ const std::vector<uint8_t> &BpfProgram::getBTF()
+      {
+        LOG(ERROR) << "Relocation in " << relsecname << " type " << reltype
+                   << " sym " << relsym << " type " << symtype;
+        throw std::invalid_argument(
+            "Unsupported symbol type referenced in relocation");
+      }
+
+      // Assume sym->st_shndx corresponds to .text, therefore symbol value is
+      // an offset from text_offset_.
+      //
+      // Relocate via direct instruction manipulation instead of the
+      // relocation entry for readibility purposes.
+      //
+      // This is a hack. We shouldn't do this. However, we don't actually have
+      // the ELF section table to determine if the relocation actually refers
+      // to .text
+      auto target_insn = (text_offset_ + sym->st_value + insn->imm) /
+                         sizeof(struct bpf_insn);
+      insn->src_reg = BPF_PSEUDO_FUNC;
+      insn->imm = (target_insn - insn_offset - 1); // jump offset
+    }
+
+    // Step 3: relocate .text, if necessary. TODO.
+    bool need_text_rels = text_offset_ > 0 &&
+                          bytecode_.find(".rel.text") != bytecode_.end();
+    if (need_text_rels)
+      throw std::logic_error("Relocations in .text are not implemented yet");
+  }
+
+  // Step 4: deal with bpf_func_infos.
+  relocateFuncInfos();
+}
+
+// Assumed structure:
+//
+// code_[0..text_offset_) - program
+// code_[text_offset_..code_.size()) - .text
+void BpfProgram::relocateFuncInfos()
+{
+  const auto *btfsec = bytecode_.find(".BTF")->second.data();
+  auto *btfhdr = reinterpret_cast<const struct btf_header *>(btfsec);
+  auto *btfstr = btfsec + btfhdr->hdr_len + btfhdr->str_off;
+
+  const auto *btfextsec = bytecode_.find(".BTF.ext")->second.data();
+
+  auto *exthdr = reinterpret_cast<const struct btf_ext_header *>(btfextsec);
+  auto *exthdr_end = btfextsec + exthdr->hdr_len;
+  auto *func_info_end = exthdr_end + exthdr->func_info_len;
+
+  auto *ptr = exthdr_end + exthdr->func_info_off;
+
+  auto func_info_rec_size = *reinterpret_cast<const __u32 *>(ptr);
+  if (sizeof(bpf_func_info) > func_info_rec_size)
+    throw std::invalid_argument("Unsupported bpf_func_info size");
+  ptr += sizeof(__u32);
+
+  // We need to find the bpf_ext_info_secs for our program section and,
+  // optionally, for .text. They're likely not in the order we need them in,
+  // so find them first, then copy things over, so we keep the invariant that
+  // the first func_info is for the function at offset 0.
+  //
+  const btf_ext_info_sec *text_funcinfo_sec = nullptr;
+  const btf_ext_info_sec *prog_funcinfo_sec = nullptr;
+
+  while (ptr < func_info_end)
+  {
+    auto *info_sec = reinterpret_cast<const struct btf_ext_info_sec *>(ptr);
+    auto *name = reinterpret_cast<const char *>(btfstr +
+                                                info_sec->sec_name_off);
+    if (text_offset_ > 0 && strcmp(name, ".text") == 0)
+      text_funcinfo_sec = reinterpret_cast<const btf_ext_info_sec *>(ptr);
```

Comment:
> I'm finding all casting makes the code harder to read

Same here but I don't see many ways to avoid that. That's the price of manipulating raw data which is the way both libelf and libbpf work.